feat: Update GitHub and Bitbucket setup flow with custom incident detection PUL-1673 (#198)
codacy/pulse-user-docs: docs/one-click-integrations/github-integration.md

1.  Choose the strategy to detect **incidents** that best fits your workflows. See the [section below](#incident-detection-strategy) for a detailed description of each option.

    !!! note

>         Pulse can only detect incidents automatically from GitHub if you configured the **deployment detection strategy** to use [merged pull requests](#gh-deploy-merged-pr).

Reviewer: Minor tweak to avoid using the past tense.

```suggestion
        Pulse can only detect incidents automatically from GitHub if you configure the **deployment detection strategy** to use [merged pull requests](#gh-deploy-merged-pr).
```

## Automatic incident detection strategies {: id="incident-detection-strategy"}

The Pulse GitHub integration can detect **incidents** automatically using the following strategies:

-   [Pull request reverts](#gh-incident-pr-revert)
-   [Pull request naming convention](#gh-incident-naming)

You can also choose [not to detect incidents via GitHub](#gh-incident-not-detect).

!!! note

>     Pulse incident detection mechanism bases on pull requests. Thus, Pulse can only detect incidents automatically from GitHub if you configured the deployment detection strategy to use [merged pull requests](#gh-deploy-merged-pr).

Reviewer: The initial sentence is a bit unclear. Here's a suggested edit (or we could also consider saying that `Pulse analyzes pull requests`):

```suggestion
    Pulse uses pull requests to detect incidents. Thus, Pulse can only detect incidents automatically from GitHub if you configured the deployment detection strategy to use [merged pull requests](#gh-deploy-merged-pr).
```

Author: Yep, much clear. Thanks!

-   Pulse considers an incident any pull request that **targets the default branch** of the repository merged from a branch whose name starts with `revert-`, getting the number of the reverted pull request from the branch name, `revert-<pull request number>`. If you change the name of the branch created by GitHub when you revert a pull request, Pulse may not be able to obtain the incident data correctly.

-   The incident creation date is the timestamp when the reverted pull request was initially merged. If Pulse can't get the reverted pull request number from the branch name, the incident creation date is the timestamp of the first commit to the incident pull request.

-   Pulse associates incidents to the system matching the repository name.

### Use pull request naming convention (based on default branch) {: id="gh-incident-naming"}

-   Pulse applies the same mechanism to detect incidents used for the [pull request reverts](#gh-incident-pr-revert) strategy. However, instead of considering branches whose name starts with `revert-`, Pulse uses the prefix that you define for the **branch name** or the **pull request title** to match the naming convention of your workflow.

>     ![Pulse custom incident detection](images/github-incident-naming-convention.png)

Reviewer 2: clean: We could omit this screenshot here and update the main screenshot `github-incident-strategy.png` to have the naming convention option selected and showing the available configurations.

Reviewer 2: See https://github.com/codacy/pulse-user-docs/pull/198#discussion_r1117385468. :heavy_check_mark:

> -   Your prefix must have 2 to 50 characters. For a **branch name**, your prefix can't contain spaces or special characters, such as `~ ^ : ? * [ ]`. See the [Git reference](https://git-scm.com/docs/git-check-ref-format) for further details.

Reviewer 3: Same as above.

Author: Thanks! Fixed! 👍

> -   Your prefix must have 2 to 50 characters. For a **branch name**, your prefix can't contain spaces or special characters, such as `~`, `^`, `:`, `?`, `*`, `[`, `]`. See the [Git reference](https://git-scm.com/docs/git-check-ref-format) for further details.

Reviewer 2: style: Simplify list of special characters.

```suggestion
-   Your prefix must have 2 to 50 characters. For a **branch name**, your prefix can't contain spaces or special characters, such as `~ ^ : ? * [ ]`. See the [Git reference](https://git-scm.com/docs/git-check-ref-format) for further details.
```

> 

Author: @afsmeira I added the prefix restrictions here. Can you please validate?

-   The incident creation date is the timestamp when the reverted pull request was initially merged. If Pulse can't get the reverted pull request number from the branch name, the incident creation date is the timestamp of the first commit to the incident pull request.
-   Pulse associates incidents to the system matching the repository name.

### Use pull request naming convention (based on default branch) {: id="gh-incident-naming"}

> -   Pulse executes the same incident detection mechanism based on [pull request reverts](#gh-incident-pr-revert) using the naming convention you define instead of the default branch prefix `revert-`.

Reviewer: I found this sentence a bit hard to read. Here's my attempt at simplifying it, let me know 😄 

```suggestion
-   Pulse can also flag incidents when it detects [pull request reverts](#gh-incident-pr-revert) using a branch prefix you define instead of the default branch prefix `revert-`.
```

Reviewer 3: I think it's a bit strange that we mention branch prefix in this sentence but in the next sentence the prefix can be for the branch name or PR title.

Reviewer: Fair point @afsmeira. @claudiacarpinteiro I'm updating my suggestion.

```suggestion
-   Pulse can also flag incidents when it detects [pull request reverts](#gh-incident-pr-revert) whose title or related branch name use a custom prefix.
```

Author: This is not exactly what I'm trying to say... I'll try to rephrase it.

@afsmeira For the custom incident detection, from where are we getting the number of the reverted pull request? Doe the rule for "Use pull request reverts" still apply?

Reviewer 3: Yes, after we match the prefix we try to get the PR number that was reverted in the same way that we do for the non-custom detection.

Author: @nicklem @afsmeira I rephrased it, please have a look. Hope it's clear now. 🙏

Reviewer: Looking good! 🚀

> -   You can define the prefix for the branch name or the pull request title to match the naming convention on your workflow.

Reviewer: I believe we can use `of` here since this could be reworded as `your workflow's naming convention`:

```suggestion
-   You can define the prefix for the branch name or the pull request title to match the naming convention of your workflow.
```
